### PR TITLE
Adding cuda kernel (optimized for sm80) for block-wise 4b quantized float 16 GEMM.

### PR DESCRIPTION
**Pull request by chenfucn**
### Description
Adding CUDA kernel for block-wise 4b quantized float 16 GEMM, this is specially optimized for Nvidia Ampere GPUs. 

### Motivation and Context
Trying to improve quantized LLM inference performance on Nvidia Ampere GPUs

### Note:
This is implemented by extending CUTLASS, so it has a hard dependency on CUTLASS. However, in current build system, loading of CUTLASS dependency is guarded with:

(onnxruntime_USE_FLASH_ATTENTION OR onnxruntime_USE_MEMORY_EFFICIENT_ATTENTION)

If both of these options are turned off, then compilation will fail.

Why CUTLASS dependency is guarded at all? It's a header file only library that does not introduce any binary if not instantiated. What's the downside of removing all the guards and just include CUTLASS unconditionally?

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -699,6 +700,10 @@ if (onnxruntime_USE_CUDA)
     set(onnxruntime_USE_FLASH_ATTENTION OFF)
     set(onnxruntime_USE_MEMORY_EFFICIENT_ATTENTION OFF)
   endif()
+  if (CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11.4)
+    message( STATUS "Turn off cutlass since CUDA compiler version < 11.6")
```
[the remark this answers is not in the document]
Comment (chenfucn) [reply cmake/CMakeLists.txt:704]: CUTLASS document says it's minimum required cuda toolkit version is 11.4

File: onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+  ORT_ENFORCE(passed, "Gemm kernel result wrong!");
+}
+
+template void run_blkq4_gemm<16, true, false, true>(int m, int n, int k);
+template void run_blkq4_gemm<16, true, false, false>(int m, int n, int k);
+template void run_blkq4_gemm<32, true, false, true>(int m, int n, int k);
+template void run_blkq4_gemm<32, true, false, false>(int m, int n, int k);
+template void run_blkq4_gemm<64, true, false, true>(int m, int n, int k);
+template void run_blkq4_gemm<64, true, false, false>(int m, int n, int k);
+template void run_blkq4_gemm<16, false, false, true>(int m, int n, int k);
```
Comment (yufenglee) [inline onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu:468]: we also need to support block size 128, even 256

File: onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+                || std::is_same<LayoutCutlass, cutlass::layout::RowMajor>::value);
+  auto shape = make_Position(tensor.extent().row(), tensor.extent().column());
+  return MatrixRef<Element const, Layout, true>(tensor.host_data(), tensor.capacity(), shape);
+}
+
+//
+// Invoking the kernel
+//
+
+template<
```
Comment (yufenglee) [inline onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu:146]: nit: the function is too long

File: onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      cutlass::TensorRef<ElementQScale const, LayoutInputQScale> ref_Qscale_,
+      cutlass::TensorRef<ElementQOffset const, LayoutInputQScale> ref_Qoffset_,
+      cutlass::TensorRef<ElementOutput const, LayoutOutput> ref_C_,
+      cutlass::TensorRef<ElementOutput, LayoutOutput> ref_D_,
+      typename EpilogueOp::Params epilogue_ = typename EpilogueOp::Params()) {
+    if constexpr (!kHasQuantOffset) {
+      return cutlass::Status::kErrorNotSupported;
+    } else {
+      if constexpr (ShapeMMAThreadBlock::kM == 16) {
+        if (problem_size_.m() > 16) {
```
Comment (yufenglee) [inline onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h:131]: the special kernel for M<=16 is yet to implement?

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+    /// Element type for B matrix operand
+    typename ElementB_,
+    /// Layout type for B matrix operand
+    typename LayoutB_,
+    /// Element type for quant scales
+    typename ElementQScale_,
+    /// Element type for quant offsets
+    typename ElementQOffset_,
+    /// Layout type for quant scales and offsets
+    typename LayoutQScale_,
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h:88]: >LayoutQScale_ [](http://example.com/codeflow?start=13&length=13)

nit: a better name or split for offset and scale

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+    return bytes;
+  }
+
+  /// Initializes GEMM state from arguments.
+  Status initialize(Arguments const &args, void *workspace = nullptr, cudaStream_t stream = nullptr) {
+
+    // Determine grid shape
+    ThreadblockSwizzle threadblock_swizzle;
+
+    cutlass::gemm::GemmCoord grid_shape = threadblock_swizzle.get_tiled_shape(
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h:362]: >grid_shape [](http://example.com/codeflow?start=29&length=10)

nit: tiled_shape

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+/*! A specialized GEMM operator for quantized B GEMM.
+
+  It is modified from cutlass::gemm::device::Gemm. Both this class and the original Gemm class
+  are pretty much boilerplate code that construct the Gemm kernel class, and pass parameters
+  and controls to it. The only difference is that this class has a few more template parameters
+  to support quantization.
+
+  This implementation pretty much follows the design of cutlass. But this class seems to be
+  just a wrapper of the Gemm kernel class. Is this really necessary?
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h:71]: >Is this really necessary? [](http://example.com/codeflow?start=43&length=25)

Gemm Kernel is a cuda kernel implementation and this class is the invocation of the kernel from host, i think

File: onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu (hunk 2 above)
Comment (chenfucn) [reply onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu:468]: currently not supported. a dispatcher function will return error in later changes.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h (hunk 6 above)
Comment (chenfucn) [reply onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h:362]: this is following the naming from CUTLASS' own gemm.h. I am trying to keep the original structure intact.

File: onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu (hunk 3 above)
Comment (chenfucn) [reply onnxruntime/test/providers/cuda/test_cases/blkq4_fp16_gemm_sm80_testcu.cu:146]: tried to break it up. I think it won't help. This function contains test data setup, run gemm under test, run reference gemm, compare results. All necessary steps for a test. If I were to break it up to multiple functions, I would need to repeat many of the type and shape definitions, probably won't help with readability.

File: onnxruntime/core/mickey/blk_q4/f16_prepack_sm80.h [hunk 8]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  *
  * Module Name:
- *    prepack_sm80.h
+ *    blk_q4/f16_prepack_sm80.h
```
Comment (yufenglee) [inline onnxruntime/core/mickey/blk_q4/f16_prepack_sm80.h:6]: for columnwise, prepacking of metadata can also avoid stride read.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/kernel/quantb_gemm.h [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+    }
+    if (qscale_k * Mma::QuantBlocking::kRow != problem_size_k) {
+      printf("qscale_k * Mma::QuantBlocking::kK != problem_size_k! can_implement() should have returned false!\n");
+    }
+    if (qscale_n == 0) {
+      printf("qscale_n is 0! can_implement() should have returned false!\n");
+    }
+    if (qscale_n * Mma::QuantBlocking::kColumn != params.problem_size.n()) {
+      printf("qscale_n * Mma::QuantBlocking::kN != params.problem_size.n()! can_implement() should have returned false!\n");
+    }
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/kernel/quantb_gemm.h:326]: nit: use assert instead

File: onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  // Layout of quantization scale and offset, oriented to be loaded using less instructions
+  // in a warp tile
+  using LayoutInputQScale =
+      typename std::conditional<QuantBlocking::kRow == 1,
+                                cutlass::layout::ColumnMajor,
+                                cutlass::layout::RowMajor>::type;  // <- layout of quantization scale
+
+  using ShapeMMAThreadBlock =
+      typename std::conditional<SmallM,
+                                cutlass::gemm::GemmShape<16, 64, 64>,
```
Comment (yufenglee) [inline onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h:66]: >64 [](http://example.com/codeflow?start=61&length=2)

Can we use 256 for SmallM case? In this way, the kernel can be extended to support block size 128 and 256. Now it is quite limited to only support block size <=64

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 11]
```diff
@@ -0,0 +1,787 @@
...
+////////////////////////////////////////////////////////////////////////////////
+
+namespace{
+
+struct b32_pair{
+  uint32_t a;
+  uint32_t b;
+};
+
+struct fp16_quard{
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:43]: >fp16_quard [](http://example.com/codeflow?start=7&length=10)

nit: fp16_quad

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 12]
```diff
@@ -0,0 +1,787 @@
...
+    // static_cast<cutlass::half_t>(16 + weight)
+    // 4b weights are prepacked into [0, 2, 4, 6, 1, 3, 5, 7], so that adjacent weights
+    // are in different 16b half words, making it easier to convert to fp16.
+    asm volatile(
+        "{\n\t"
+        "  shl.b32       %0, %4, 6;\n"
+        "  shl.b32       %1, %4, 2;\n"
+        "  shr.u32       %2, %4, 2;\n"
+        "  shr.u32       %3, %4, 6;\n"
+        "  lop3.b32      %0, %0, 0x03c003c0, 0x4c004c00, 0xea;\n" // a & 0x03c0 | 0xcc00
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:91]: >0xcc00 [](http://example.com/codeflow?start=82&length=6)

typo: 0x4c00

File: onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h (hunk 10 above)
Comment (chenfucn) [reply onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h:66]: These numbers are based on performance measurements. Changing it to 256 would slow it down significantly.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h (hunk 7 above)
Comment (chenfucn) [reply onnxruntime/core/mickey/cutlass_ext/q4gemm/device/quantb_gemm.h:71]: This is my first cuda adventure and CUTLASS offered a good starting point for me. Looking back, however, I think some of these higher-level components maybe unnecessary, even somewhat restrictive (as a training wheel). Going forward, we should study how to use CUTLASS lower-level components in a more flexible way.

File: onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h (hunk 10 above)
Comment (chenfucn) [reply onnxruntime/core/mickey/blk_q4/f16_gemm_sm80.h:66]: for column-wise block quantization, relaxing N stride does not help with supporting more block size. relaxing K-stride would require rewrite memory swizzling logic. I do plan to rewrite the memory swizzling part in the next step of performance optimization of cuda kernels.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 13]
```diff
@@ -0,0 +1,787 @@
...
+  CUTLASS_HOST_DEVICE
+  void load(FragmentScale &frag, FragmentOffset &frag_offset) {
+    if constexpr(kNumBsPerCoreTileFragement == 2
+                 && kBTilesPerMma == 2
+                 && BlockingShape::kRow == 1){
+      // Optimize for a special case of:
+      //    16b gemm (kNumBsPerCoreTileFragement == 2)
+      //    2 B operand tiles per mma (kBTilesPerMma == 2)
+      //    (1,n) quantization blocking
+      // The weight and offset tensor is prepacked to reduce load instructions.
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:359]: >weight [](http://example.com/codeflow?start=13&length=6)

typo: scale

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 14]
```diff
@@ -0,0 +1,787 @@
...
+#else
+          assert(0);
+#endif
+
+          offsets_ptr += 4;
+        } else {
+          offsets.fp16_quad.a = scales_ptr->fp16_quad.a * static_cast<cutlass::half_t>(-16-8);
+          offsets.fp16_quad.b = scales_ptr->fp16_quad.b * static_cast<cutlass::half_t>(-16-8);
+          offsets.fp16_quad.c = scales_ptr->fp16_quad.c * static_cast<cutlass::half_t>(-16-8);
+          offsets.fp16_quad.d = scales_ptr->fp16_quad.d * static_cast<cutlass::half_t>(-16-8);
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:493]: use fp16x2 mul to save instructions.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 15]
```diff
@@ -0,0 +1,787 @@
...
+#else
+          assert(0);
+#endif
+          dest_pair += 2;
+        }
+        scales_ptr++;
+      }
+
+    } else {
+      // unoptiomized path for other cases, very slow
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:516]: >// unoptiomized path for other cases, very slow [](http://example.com/codeflow?start=6&length=47)

columnwise quantization is what we are heavily using now.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_mma_tensor_op.h [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+namespace cutlass {
+namespace gemm {
+namespace warp {
+
+/////////////////////////////////////////////////////////////////////////////////////////////////
+
+namespace internal {
+
+template <typename T, typename S, int N, FloatRoundStyle Round>
+struct ConvertAndPack {
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_mma_tensor_op.h:74]: >ConvertAndPack [](http://example.com/codeflow?start=7&length=14)

where is it used?

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 17]
```diff
@@ -0,0 +1,787 @@
...
+  CUTLASS_DEVICE
+  static MatrixCoord lane_position(int lane_id) {
+    if constexpr(kNumBsPerCoreTileFragement == 2
+                 && kBTilesPerMma == 2
+                 && BlockingShape::kRow == 1){
+      // Optimize for a special case of:
+      //    16b gemm (kNumBsPerCoreTileFragement == 2)
+      //    2 B operand tiles per mma (kBTilesPerMma == 2)
+      //    (1,n) quantization blocking
+      // The weight and offset tensor is prepacked to reduce load instructions.
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:200]: >weight [](http://example.com/codeflow?start=13&length=6)

typo: scale

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 18]
```diff
@@ -0,0 +1,787 @@
...
+    // This is for operand B, so advance on the K dimension
+    lane_position_ += make_Coord(MetaTile::TileShapeB::kRow, 0);
+    return *this;
+  }
+
+  /// Advances the pointer
+  CUTLASS_HOST_DEVICE
+  QuantBMetaMmaTensorOpTileIterator &operator--() {
+    // This is for operand B, so advance on the K dimension
+    lane_position_ += make_Coord(MetaTile::TileShapeB::kRow, 0);
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:557]: >+= [](http://example.com/codeflow?start=19&length=2)

-=?

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h [hunk 19]
```diff
@@ -0,0 +1,787 @@
...
```
Comment (yufenglee) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:768]: >+= [](http://example.com/codeflow?start=19&length=2)

should it be -=?

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h (hunk 15 above)
Comment (chenfucn) [reply onnxruntime/core/mickey/cutlass_ext/q4gemm/warp/quantb_meta_mma_tensor_op_tile_iterator.h:529]: That part is not implemented here. It's in the other specialization of this template. The shape of the tensor core tile, and the layout of the meta tensor, and the way tile iterator works (InstructionShape::kK, WarpShape::kN), really does not leave a lot of room for optimization there.

**Pull request merged**